addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -192,105 +231,168 @@ func _after_tick(_delta, _tick):
-		pass
-	
-	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state >= 0:
+func _submit_state(received_state: Dictionary, tick: int):
+	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state_tick >= 0:
 		# State too old!
 		_logger.error("Received state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
 		return
-
-	var sender = multiplayer.get_remote_sender_id()
-	var sanitized = {}
-	for property in state:
-		var pe = _property_cache.get_entry(property)
-		var value = state[property]
-		var state_owner = pe.node.get_multiplayer_authority()
 		
-		if state_owner != sender:
+	#if (_states.has(tick)):
+		#_logger.warning("Received state %s AGAIN" % tick)
+		#return
+
+	# It is guaranteed that the first states are full, until the client sends acks to the server
+	# which enables the server to start sending diffs
+	if (not _has_received_full_state):
+		_has_received_full_state = true
+		_receive_full_state_ack.rpc_id(1, tick)
+
+	if (received_state.size() > 0):
+		set_received_state(received_state, tick, multiplayer.get_remote_sender_id())
+	elif (NetworkRollback.enable_state_diffs):
+		if (_states.has(tick - 1)):
+			_states[tick] = _states[tick - 1]
+			_latest_state_tick = tick
+		else: #packet loss or unreliable_ordered didn't work so we got future tick before previous tick (e.g. received 106 diff which is empty, but we dont have 105 which is also empty)
+			_logger.warning("non-linear diff states. Received tick %s while latest tick is %s" % [tick, _latest_state_tick])
+			if (_states.is_empty() == false && tick > _latest_state_tick):
+				for picked_tick in range(_latest_state_tick, tick + 1):
+					_states[tick] = _states[_latest_state_tick]
+				_latest_state_tick = tick
+			else:
+				_logger.error("Received a state without any properties from %s, but missing previous states!" % multiplayer.get_remote_sender_id())
+				breakpoint
+
+
+	else:
+		_logger.error("Received a state without any properties from %s, but diff states which should make it possible, are disabled!" % multiplayer.get_remote_sender_id())
+		breakpoint
+
+func set_received_state(received_state: Dictionary, tick: int, sender_id: int):
+	var sanitized: Dictionary = {}
+	var missing_property: bool = false
+	for property_entry in _record_state_props:
+		var state_owner_id: int = property_entry.node.get_multiplayer_authority()
+		if state_owner_id != sender_id:
 			_logger.warning("Received state for node owned by %s from %s, sender has no authority!" \
-				% [state_owner, sender])
+				% [state_owner_id, sender_id])
 			continue
-		
-		sanitized[property] = value
-	
+
+		if (received_state.has(property_entry._path) == false):
+			missing_property = true
+			continue
+
+		var value = received_state[property_entry._path]
+		sanitized[property_entry._path] = value
+
+	## Detect if new property is added on runtime and sent
```

Comment:
>Why detect a new property if we then don't use the result? 

I planned to have this as a seperate PR, because it needs a lot of testing to confirm a brand new property works out all right. At least there are 2 codeblocks which address adding a new property in runtime, one here which can be expanded, and one in `_record_tick`

Reply:
Thinking about it, at `_record_tick` if a new property is added, it should simply send a full state for that tick. So there is no need to check for new properties inside `_submit_state`.